### ISSUE TITLE
学習並列化とエポックの定義変更

### ISSUE THREAD

**Issue by YuriCat** [YuriCat/FujiGokoroUECda#4]
パラメータ数が増えたので、学習前の特徴量解析+ パラメータ更新を並列化しないと遅い

あと棋譜数が増えたので、データ全体を見るのを1エポックとして制御する必要はないのでそこも柔軟に変更する

**Comment by YuriCat** [comment]
PolicyからLearnerへのポインタを辿っている現在の実装では1つのPolicyに複数のLearnerを割当てられない（配列にして強引に割り当ててもいいけど）

それよりLearnerがPolicyへのポインタを持っておき、Policyの代わりを務められるようにしておくほうがいいのでは(1つのPolicyを複数のLearnerが参照すれば並列に学習できる)

**Comment by YuriCat** [comment]
とりあえず並列で特徴量解析と学習が回るようにはなったが、
iteration 20 change trials 548992640 play trials 841749026
Change Training : MHR(BHR) = 0.490036(0.508219), KLD = 6.37549, ENT = 0.538497
Change Training : MHR(BHR) = 0.55667(0.564861), KLD = 5.80908, ENT = 0.292398
Play   Training : MHR(BHR) = 0.618106(0.679079), KLD = 1.52816, ENT = 0.767664
Change Test     : MHR(BHR) = 0.588556(0.611049), KLD = 2.87751, ENT = 0.591893
Change Test     : MHR(BHR) = 0.627179(0.648707), KLD = 2.01339, ENT = 0.516135
Play   Test     : MHR(BHR) = 0.626613(0.713206), KLD = 1.03351, ENT = 0.93825
KLやエントロピーの表示がおかしいのか、それともパラメータ更新が交互に起こるのでこうならざるを得無いのか

**Comment by YuriCat** [comment]
バッチサイズ1で逐次更新にするのが短時間での性能的にはよさそう。
そのときに正則化どうするかが問題。

learning rate = 5e-06
attenuation rate = 5e-11
L1 reguralization rate = 0
L2 reguralization rate = 1e-09
batch size = 1
0sec. iteration 0 change trials 0 play trials 0
Change Training : MHR(BHR) = 0.491768(0.628621), KLD = 1.63534, ENT = 1.71671
Change Training : MHR(BHR) = 0.557475(0.665744), KLD = 1.16366, ENT = 1.22221
Play   Training : MHR(BHR) = 0.621049(0.716381), KLD = 1.00798, ENT = 0.989973
Change Test     : MHR(BHR) = 0.506271(0.635494), KLD = 1.59272, ENT = 1.61434
Change Test     : MHR(BHR) = 0.568846(0.671214), KLD = 1.13853, ENT = 1.15983
Play   Test     : MHR(BHR) = 0.621851(0.716793), KLD = 1.00681, ENT = 0.986369
1201sec. iteration 1 change trials 30291232 play trials 463799741
Change Training : MHR(BHR) = 0.507367(0.635805), KLD = 1.58668, ENT = 1.60423
Change Training : MHR(BHR) = 0.569297(0.670852), KLD = 1.14077, ENT = 1.15133
Change Test     : MHR(BHR) = 0.507999(0.636479), KLD = 1.58504, ENT = 1.60016
Change Test     : MHR(BHR) = 0.570951(0.671776), KLD = 1.13647, ENT = 1.14442
1610sec. iteration 2 change trials 60582464 play trials 463799741
Change Training : MHR(BHR) = 0.508581(0.636199), KLD = 1.58252, ENT = 1.59391
Change Training : MHR(BHR) = 0.570801(0.671096), KLD = 1.14006, ENT = 1.13963
Change Test     : MHR(BHR) = 0.508169(0.63661), KLD = 1.58307, ENT = 1.59707
Change Test     : MHR(BHR) = 0.570901(0.672059), KLD = 1.13652, ENT = 1.14196
2019sec. iteration 3 change trials 90873696 play trials 463799741
Change Training : MHR(BHR) = 0.508898(0.63635), KLD = 1.5808, ENT = 1.59115
Change Training : MHR(BHR) = 0.571399(0.67116), KLD = 1.13993, ENT = 1.13463
Change Test     : MHR(BHR) = 0.507457(0.636467), KLD = 1.58155, ENT = 1.60288
Change Test     : MHR(BHR) = 0.571998(0.671627), KLD = 1.13623, ENT = 1.13558
2425sec. iteration 4 change trials 121164928 play trials 463799741
Change Training : MHR(BHR) = 0.508991(0.636352), KLD = 1.57986, ENT = 1.59016
Change Training : MHR(BHR) = 0.571737(0.671168), KLD = 1.13984, ENT = 1.13169
Change Test     : MHR(BHR) = 0.509268(0.636379), KLD = 1.58068, ENT = 1.58937
Change Test     : MHR(BHR) = 0.572884(0.671675), KLD = 1.13629, ENT = 1.12802
2833sec. iteration 5 change trials 151456160 play trials 463799741
Change Training : MHR(BHR) = 0.509021(0.636423), KLD = 1.57925, ENT = 1.58994
Change Training : MHR(BHR) = 0.571969(0.671146), KLD = 1.1398, ENT = 1.12974
Play   Training : MHR(BHR) = 0.622139(0.71697), KLD = 1.00598, ENT = 0.985406
Change Test     : MHR(BHR) = 0.509686(0.636077), KLD = 1.58048, ENT = 1.58555
Change Test     : MHR(BHR) = 0.573387(0.671694), KLD = 1.1358, ENT = 1.1245
Play   Test     : MHR(BHR) = 0.622131(0.716878), KLD = 1.00663, ENT = 0.985655

**Comment by YuriCat** [comment]
正則化はいびつだが、合法手の数に依存する形で拘束の強さが決まるように入れている
エポックの定義はそのままだがとりあえず問題なさそうなので閉じる
